src/cosalib/aws.py

```diff
@@ -54,37 +86,47 @@ def aws_run_ore_replicate(build, args):
 
     source_image = buildmeta['amis'][0]['hvm']
     source_region = buildmeta['amis'][0]['name']
-    ore_args = ['ore']
-    if args.log_level:
-        ore_args.extend(['--log-level', args.log_level])
-    ore_args.extend([
-        'aws', 'copy-image', '--image',
-        source_image, '--region', source_region
-    ])
 
-    upload_failed_in_region = None
-    for upload_region in region_list:
-        region_ore_args = ore_args.copy() + [upload_region]
-        print("+ {}".format(subprocess.list2cmdline(region_ore_args)))
-        try:
-            ore_data = json.loads(subprocess.check_output(region_ore_args))
-        except subprocess.CalledProcessError:
-            upload_failed_in_region = upload_region
-            break
-        # This matches the Container Linux schema:
-        # https://stable.release.core-os.net/amd64-usr/current/coreos_production_ami_all.json
-        ami_data = [{'name': name,
-                     'hvm': vals['ami'],
-                     'snapshot': vals['snapshot']}
-                    for name, vals in ore_data.items()]
-        buildmeta['amis'].extend(ami_data)
-        # Record the AMI's that have been replicated as they happen.
-        # When re-running the replication, we don't want to be lose
-        # what has been done.
-        build.meta_write()
-
-    if upload_failed_in_region is not None:
-        raise Exception(f"Upload failed in {upload_failed_in_region} region")
+    # Create a shared dict for writing values
+    # and passing values. We cannot put the build.meta['amis']
+    # here since the Manager does allow mutating nested lists.
+    manager = Manager()
+    shared_meta = manager.dict({
+        "source_region": source_region,
+        "source_image": source_image,
+        "args": args,
+    })
+
+    # Create the workers to do the work.
+    workers = []
+    for region in region_list:
+        log.info(f"starting ore worker for {region}")
+        p = Process(target=_replicate_worker, args=(region, shared_meta))
```

Comment:
See also 56952ffae17baf542f13563a2a4ae0e45ffeccfb - the Python multiprocessing default of `fork()` *will* cause evil bugs.